### PR TITLE
Dynamic vars

### PR DESCRIPTION
**Pull request by marcelocantos**
Fixes # .

Changes proposed in this pull request:
- Implement dynamically scoped names, which remain in scope in every stack frame under the point at which the name was bound.

Example:

```sh
arrai eval '
let f = \x ${prefix} ++ x;
[
	let ${prefix} = "FOO: "; f("hello"),
	let ${prefix} = "BAR: "; f("goodbye"),
]'
['FOO: hello', 'BAR: goodbye']
```

When `f` is called, it sees different values of `${prefix}`, as bound by each `let` form.

Checklist:
- [x] Added related tests
- [x] Made corresponding changes to the documentation

### CODE REVIEW

File: rel/expr_tuplemap.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ func (e *TupleMapExpr) Eval(ctx context.Context, local Scope) (_ Value, err erro
 		}
 	}()
 	return value.(Tuple).Map(func(v Value) Value {
-		scope, _ := e.fn.arg.Bind(ctx, local, v) //nolint: errcheck
+		ctx, scope, _ := e.fn.arg.Bind(ctx, local, v) //nolint: errcheck
```
Comment (nofun97) [inline rel/expr_tuplemap.go:41]: why isn't error handled?

File: syntax/expr_import.go [hunk 2]
```diff
@@ -15,7 +15,7 @@ type ImportExpr struct {
 }
 
 func NewImportExpr(scanner parser.Scanner, imported rel.Expr, path string) ImportExpr {
-	return ImportExpr{rel.ExprScanner{scanner}, NewPackageExpr(scanner, imported), path}
+	return ImportExpr{rel.ExprScanner{Src: scanner}, NewPackageExpr(scanner, imported), path}
```
Comment (nofun97) [inline syntax/expr_import.go:18]: seems like an unnecessary change

File: rel/pattern_set.go [hunk 3]
```diff
@@ -102,7 +91,11 @@ func (p SetPattern) Bind(ctx context.Context, local Scope, value Value) (context
 				return ctx, EmptyScope, fmt.Errorf("the length of set %s is wrong", set)
 			}
 
-			ctx, scope, err = p.patterns[i].Bind(ctx, local, set.set.Any().(Value))
+			e := set.Enumerator()
+			if !e.MoveNext() {
+				panic("set with count 1 failed to enumerate")
```
Comment (nofun97) [inline rel/pattern_set.go:96]: is this an implementation error? if not, it should return the error.

**Review by nofun97 — COMMENTED** [review]
Some comments, more will come

File: docs/README.md [hunk 4]
```diff
@@ -408,6 +408,48 @@ If `\name` is omitted, `\.` is assumed.
...
+};
+(
+   round: round,
+   add: \x \y round(x + y),
+   sub: \x \y round(x - y),
+   mul: \x \y round(x * y),
+   div: \x \y round(x / y),
+)
+
+# app.arrai
```
Comment (nofun97) [inline docs/README.md:439]: add expected output for `app.arrai`

**Review by nofun97 — CHANGES_REQUESTED** [review]

File: rel/pattern_set.go (hunk 3 above)
Comment (marcelocantos) [reply rel/pattern_set.go:96]: It is an implementation error. The count is checked a few lines prior.

File: syntax/expr_import.go (hunk 2 above)
Comment (marcelocantos) [reply syntax/expr_import.go]: golangci-lint disagrees

File: docs/README.md (hunk 4 above)
Comment (marcelocantos) [reply docs/README.md:439]: Done

File: rel/expr_tuplemap.go (hunk 1 above)
Comment (marcelocantos) [reply rel/expr_tuplemap.go]: It is now

File: rel/expr_tuplemap.go [hunk 5]
```diff
@@ -31,14 +31,11 @@ func (e *TupleMapExpr) Eval(ctx context.Context, local Scope) (Value, error) {
 		return nil, WrapContextErr(err, e, local)
 	}
 	return value.(Tuple).Map(func(v Value) (Value, error) {
-		scope, err := e.fn.arg.Bind(ctx, local, v) //nolint: errcheck
+		ctx, scope, err := e.fn.arg.Bind(ctx, local, v) //nolint: errcheck
```
Comment (nofun97) [inline rel/expr_tuplemap.go:34]:
```suggestion
		ctx, scope, err := e.fn.arg.Bind(ctx, local, v)
```

File: syntax/expr_import.go [hunk 6]
```diff
@@ -15,7 +15,11 @@ type ImportExpr struct {
 }
 
 func NewImportExpr(scanner parser.Scanner, imported rel.Expr, path string) ImportExpr {
-	return ImportExpr{rel.ExprScanner{Src: scanner}, NewPackageExpr(scanner, imported), path}
+	return ImportExpr{
+		ExprScanner: rel.ExprScanner{Src: scanner},
```
Comment (nofun97) [inline syntax/expr_import.go]: nit:
```suggestion
		ExprScanner: rel.ExprScanner{scanner},
```

**Review by nofun97 — APPROVED** [review]
LGTM

File: rel/pattern_set.go [hunk 7]
```diff
@@ -22,37 +22,25 @@ func NewSetPattern(patterns ...Pattern) SetPattern {
 	return SetPattern{patterns}
 }
 
-func (p SetPattern) Bind(ctx context.Context, local Scope, value Value) (Scope, error) {
-	set, is := value.(GenericSet)
-	if !is {
-		return EmptyScope, fmt.Errorf("value %s is not a set", value)
```
Comment (orlade-anz) [inline rel/pattern_set.go:28]: Boo :(
